sys/net/gnrc/netreg/gnrc_netreg.c

```diff
@@ -33,8 +33,12 @@
 /* The registry as lookup table by gnrc_nettype_t */
 static gnrc_netreg_entry_t *netreg[GNRC_NETTYPE_NUMOF];
 
+mutex_t gnrc_netreg_lock;
+
 void gnrc_netreg_init(void)
 {
+    mutex_init(&gnrc_netreg_lock);
```

Comment:
From a compactness point of view, I'd guess this is smaller (one fewer thing in .data). From a safety point of view (if we do want to catch things around netreg stuff being done before init), I could even move this to MUTEX_INIT_LOCKED and only unlock in the init function ... yes, I think I'd like that best.

Reply:
It would not keep it idempotent -- is that a desirable property? I'd think that while init functions would practically be idempotent, calling them multiple times would be an error that's better detected than silently accepted.

Reply:
Especially with mutexes, if `init` is not idempotent, the unittests might fail ;-)

Reply:
This is trusting that the re-init catches all effect that were had on other systems -- but OK, if it works, I won't break it :-)

But the harder question is really that of whether we have a good RW lock to switch to -- a) because I don't find all the numof calls that'd break due to the lock-is-held assertion, but primarily b) because a GNRC component might use callbacks, the lock is obviously held across the callback, and during that callback that component would iterate over its own recipients and forward the message right away.

Would using an p_thread/rwlock be OK here, or should we rather do a very simple local one? I'm leaning toward the latter, but that might be just because I'm not too big of a fan of using POSIXish stuff on embedded systems.

Reply:
> Would using an p_thread/rwlock be OK here, or should we rather do a very simple local one?

No, definitely not use the POSIX rwlock here. Rather, if we need it here, a simple implementation would be needed. We could, if we want to, then use it with the POSIX implementation, too.

Reply:
Having looked at the pthread_rwlock in more detail, yes, this is *definitely* not something to pull into GNRC -- its complexity might be almost minimal for what it does, but that it supports locks held until some 64-bit timestamp (pulling ZTIMER64) makes it a clear no-go.

I have a hunch that one could implement a rather cheap atomics based rw-lock (not even incurring a irq_off in regular operation) if one is willing to priorize reads over writes (which I think is a practical priorization here).

I'll play, but it may take some time.

---

Independent of that: Is locking here architecturally OK even?

Requiring that a lock (mutex or rwlock) is held while a packet is dispatched necessitates that the first source of the packet is a thread, for an interrupt can't lock a mutex. May packets be dispatched from interrupts? If so, I'd need to revisit the strategy (and build a suitable synchronization primitive for that).

Reply:
In general, yes, it looks ok. However, I have some concerns as well regarding your second concern in https://github.com/RIOT-OS/RIOT/pull/18696#issuecomment-1268724567 as well, and I am not sure if a rwlock would solve that really :-/. If worse comes to worse, we just need to deactivate IRQ interrupts for netreg accesses :-/.

Reply:
Mostly to help myself thinking, I'm summarizing the locking requirements:

1. _register and _unregister write into a linked list that can be read by anyone. Read-while-write is UB (and could practically have effects like "half the pointer is updated, the other still contains the old value"). Manipulating that LL either requires some mutual exclusion with iteration of that list, or both readers and writers to use atomics (which is probably not too costly because reading can be with "relaxed" semantics).
2. Insertion is unproblematic from the memory access PoV as the insertion may be ordered before or after the iterator passes over (as long as the new item's "next" pointer is set before the item is inserted).
3. Removal needs consideration for two aspects:
    i. A reader iterating over the item needs to retain the ability to iterate -- otherwise the iteration would stop prematurely if the iterator is just at the item being removed.
    ii. When the cursor is at an item, it may *act* on that item (read more of its data items and perform callbacks) -- removal of that item needs to block because it (and data referenced through it) is being acted on. (It is acceptable to over-block on the removal).

    Item 3.i can be subsumed in this, because reading the "next" pointer is one of these accesses.

    This is the only place where there needs to be blocking -- everything else can probably be well-ordered without ever blocking.

---

Maybe it helps to split the problem into two parts:
* Keeping the list sound
* Awaiting freedom to remove an item for good

These can be well separated: In the first phase of unregistering, the LL would be bent to not point to the to-be-removed item. That's a hard thing on its own, but maybe the pot is where it's easier to boil that particular pot of water.